File: АиСД/s01e01/merge_sort.py


a = [1,3,5,0,10,-1]

def merge(a, b):
    i = 0
    j = 0
    c = [0] * (len(a) + len(b))
    while i < len(a) or j < len(b):
        if j == len(b) or (i < len(a) and a[i] < b[j]):
            c[i + j] =  a[i]
            i += 1
        else:
            c[i + j] = b[j]
            j += 1
    return c
    
def sort(a):
    n = len(a)
    if n <= 1:
        return a
    pivot = len(a) // 2
    l = sort(a[:pivot])
    r = sort(a[pivot:])
    return merge(l, r)

print(sort(a))

File: АиСД/s01e01/merge_sort.c
#include <stdio.h>

void merge(int array[], int ptr_left, int ptr_middle, int ptr_right) {
     int size_left = ptr_middle - ptr_left + 1;
     int size_right = ptr_right - ptr_middle;
     
     int left[size_left];
     int right[size_right];
     
     for (int i = 0; i < size_left; ++i) {
         left[i] = array[ptr_left + i];
     }
     for (int j = 0; j < size_right; ++j) {
         right[j] = array[ptr_middle + j + 1];
     }
     
     int i = 0;
     int j = 0;
     int k = ptr_left;
     
     while (i < size_left && j < size_right) {
         if (left[i] <= right[j]) {
             array[k++] = left[i++];
         } else {
             array[k++] = right[j++];
         }
     }
     while (i < size_left) {
         array[k++] = left[i++];
     }
     while (j < size_right) {
         array[k++] = right[j++];
     }
}

void sort(int array[], int left, int right) {
    if (left < right) {
        int ptr_middle = left + (right - left) / 2;
        sort(array, left, ptr_middle);
        sort(array, ptr_middle + 1, right);
        merge(array, left, ptr_middle, right);
    }
}


int main() {
    int array[6] = {5,4,3,0,2,1};
    
    int ptr_right = sizeof(array) / sizeof(array[0]) - 1;
    int ptr_left = 0;
    
    sort(array, ptr_left, ptr_right);
    for (int i = 0; i < ptr_right + 1; ++i) {
        printf("%d ", array[i]);
    }
    return 0;
}
